### PR TITLE
added rating for community

### PR DESCRIPTION
**Pull request by Freitag-fri**

### CODE REVIEW

File: contracts/Peeranha.sol [hunk 1]
```diff
@@ -49,9 +50,18 @@ contract Peeranha is IPeeranha, Initializable {
         SecurityLib.setupRole(userContext, SecurityLib.DEFAULT_ADMIN_ROLE, msg.sender);
     }
 
-    function getRatingToReward(address user, uint16 rewardPeriod) external view override returns(int32) {
-        RewardLib.PeriodRating storage userPeriod = RewardLib.getUserPeriodRating(userContext.userRewards, user, rewardPeriod);
-        return userPeriod.ratingToReward;
+    function getActiveInCommunity(address user, uint16 rewardPeriod) external override returns(uint32[] memory) {
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: `getActiveInCommunity` -> `getUserRewardCommunities`

Change this function to `view`

File: contracts/Peeranha.sol [hunk 2]
```diff
@@ -49,9 +50,18 @@ contract Peeranha is IPeeranha, Initializable {
...
+        require(
+            !userContext.userRatingCollection.communityRatingForUser[user].userRewards[rewardPeriod].isPaid,
+            "You already picked up this reward."
+        );
+        require(
+            userContext.userRatingCollection.communityRatingForUser[user].userRewards[rewardPeriod].activeInCommunity.length > 0,
+            "No reward for you in this period"
+        );
+
+        userContext.userRatingCollection.communityRatingForUser[user].userRewards[rewardPeriod].isPaid = true;
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: This function has to be view only and can't change any values.

In general, it is very confusing when action with name that starts with `get` changes any state.

File: contracts/PeeranhaToken.sol [hunk 3]
```diff
@@ -43,15 +42,17 @@ contract PeeranhaToken is ERC20Upgradeable, ERC20PausableUpgradeable, ERC20Cappe
    * - must be a period less then now.
   */
   function claimReward(uint16 period) external {
-    RewardLib.StatusReward storage rewardStatus = RewardLib.getRewardStatus(statusRewards, msg.sender, period);
```
Comment (sergeyi23) [inline contracts/PeeranhaToken.sol:46]: Why this was removed?
We still should use `rewardStatus` to track paid out rewards.
Remove `isPaid` from `UserRewards`.

File: contracts/libraries/RewardLib.sol [hunk 4]
```diff
@@ -18,28 +18,14 @@ library RewardLib {
...
-    mapping(address => mapping(uint16 => PeriodRating)) userRewards;
-  }
-
-  /// @notice Get information about the user's reward
-  /// @param userRewards The mapping containing all user's rewards
-  /// @param user The use who has rewards
-  /// @param period The period which we get reward
-  function getUserPeriodRating(RewardLib.UserRewards storage userRewards, address user, uint16 period) internal view returns (RewardLib.PeriodRating storage) {
-    return userRewards.userRewards[user][period];
+    uint32[] activeInCommunity;
```
Comment (sergeyi23) [inline contracts/libraries/RewardLib.sol]: `activeInCommunity` -> `rewardCommunities`

File: contracts/libraries/SecurityLib.sol [hunk 5]
```diff
@@ -195,12 +196,12 @@ library SecurityLib {
       require(false, "Action not allowed");
     }
 
-    require(user.rating >= ratingAllowen, message);
-    reduceEnergy(user, energy);
+    require(userRating >= ratingAllowen, message);
```
Comment (sergeyi23) [inline contracts/libraries/SecurityLib.sol]: Everywhere rename `ratingAllowen` -> `ratingAllowed`

File: contracts/libraries/UserLib.sol [hunk 6]
```diff
@@ -18,19 +18,33 @@ library UserLib {
...
     bytes32[] roles;
   }
 
+  struct UserRatingCollection {
+    mapping(address => CommunityRatingForUser) communityRatingForUser;
+  }
+
+  struct CommunityRatingForUser {
+    mapping(uint32 => UserRating) userRating;   //uint32 - community id
+    uint16[] rewardPeriods; // периоды когда было изменение рейтинга
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol]: Translate comment to english

File: contracts/libraries/UserLib.sol [hunk 7]
```diff
@@ -177,66 +191,87 @@ library UserLib {
...
-    uint256 pastPeriodsCount = user.rewardPeriods.length;
     
-    RewardLib.PeriodRating storage currentWeekRating = RewardLib.getUserPeriodRating(userContext.userRewards, userAddr, currentPeriod);
-    bool isFirstTransactionOnThisWeek = pastPeriodsCount == 0 || user.rewardPeriods[pastPeriodsCount - 1] != currentPeriod; 
+    CommunityRatingForUser storage communityUser = userContext.userRatingCollection.communityRatingForUser[userAddr];
+    UserRating storage userRating = communityUser.userRating[communityId];
+    if (!userRating.isActive) {
+      userRating.rating = START_USER_RATING;
+      userRating.payOutRating = START_USER_RATING;
+      userRating.isActive = true;
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol:236]: Call `communityUser.userRewards[currentPeriod].activeInCommunity.push(communityId);` here.
No need to do any checks in the loop.

File: contracts/libraries/UserLib.sol [hunk 8]
```diff
@@ -177,66 +191,87 @@ library UserLib {
...
+    int32 newRating = userRating.rating += rating;
+    uint256 pastPeriodsCount = communityUser.rewardPeriods.length;
+
+    RewardLib.PeriodRating storage currentWeekRating = RewardLib.getUserPeriodRating(communityUser.userRewards[currentPeriod], communityId);
+    bool isFirstTransactionOnThisWeek = pastPeriodsCount == 0 || communityUser.rewardPeriods[pastPeriodsCount - 1] != currentPeriod; 
     if (isFirstTransactionOnThisWeek) {
-      user.rewardPeriods.push(currentPeriod);
+      communityUser.rewardPeriods.push(currentPeriod);
+
+      bool isActiveInCommunity;  // name
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol:247]: Remove this loop. See above.

Comment (sergeyi23) [inline contracts/libraries/UserLib.sol]: I don't like that we have to read so many values every time when updating the rating for the user.

File: contracts/libraries/RewardLib.sol [hunk 9]
```diff
@@ -18,28 +18,14 @@ library RewardLib {
...
 
-  struct StatusRewardContainer {
-    mapping(address => mapping(uint16 => StatusReward)) statusReward;
-  }
-
-  struct StatusReward {
-    bool isPaid;
-  }
-
   struct UserRewards {
```
Comment (sergeyi23) [inline contracts/libraries/RewardLib.sol]: I think it should be renamed to make it clear that this are rewards for just a single period.

`UserRewards` -> `UserPeriodRewards`

File: contracts/libraries/RewardLib.sol [hunk 10]
```diff
@@ -18,28 +18,14 @@ library RewardLib {
...
-  }
-
-  /// @notice Get information about the user's reward
-  /// @param userRewards The mapping containing all user's rewards
-  /// @param user The use who has rewards
-  /// @param period The period which we get reward
-  function getUserPeriodRating(RewardLib.UserRewards storage userRewards, address user, uint16 period) internal view returns (RewardLib.PeriodRating storage) {
-    return userRewards.userRewards[user][period];
+    uint32[] activeInCommunity;
+    bool isPaid;
```
Comment (sergeyi23) [inline contracts/libraries/RewardLib.sol]: Remove from here. Return back `StatusReward`. `isPaid` has to be updated on the side of token contract.

File: contracts/libraries/RewardLib.sol [hunk 11]
```diff
@@ -18,28 +18,14 @@ library RewardLib {
     int32 ratingToReward;
   }
 
-  struct StatusRewardContainer {
```
Comment (sergeyi23) [inline contracts/libraries/RewardLib.sol:21]: Move `StatusRewardContainer` and `StatusReward` to separate lib to separate logic of main contract and token.

File: contracts/libraries/UserLib.sol [hunk 12]
```diff
@@ -177,66 +191,87 @@ library UserLib {
...
+    if (!userRating.isActive) {
+      userRating.rating = START_USER_RATING;
+      userRating.payOutRating = START_USER_RATING;
+      userRating.isActive = true;
+    }
+
+    int32 newRating = userRating.rating += rating;
+    uint256 pastPeriodsCount = communityUser.rewardPeriods.length;
+
+    RewardLib.PeriodRating storage currentWeekRating = RewardLib.getUserPeriodRating(communityUser.userRewards[currentPeriod], communityId);
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol]: is function `getUserPeriodRating` really needed now?
It does very simple thing that is not reused much.

create `storage` variable for `communityUser.userRewards[currentPeriod]`.

File: contracts/libraries/UserLib.sol (hunk 8 above)
Comment (sergeyi23) [reply contracts/libraries/UserLib.sol]: Should it check if the community needs to be added to the array even if `isFirstTransactionOnThisWeek` is `false`?

File: contracts/libraries/UserLib.sol [hunk 13]
```diff
@@ -177,66 +191,87 @@ library UserLib {
...
+    bool isFirstTransactionOnThisWeek = pastPeriodsCount == 0 || communityUser.rewardPeriods[pastPeriodsCount - 1] != currentPeriod; 
     if (isFirstTransactionOnThisWeek) {
-      user.rewardPeriods.push(currentPeriod);
+      communityUser.rewardPeriods.push(currentPeriod);
+
+      bool isActiveInCommunity;  // name
+      for (uint32 i = 0; i < communityUser.userRewards[currentPeriod].activeInCommunity.length; i++) {
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol]: What do you think if we add a field `active` to `PeriodRating` struct?

Then we could use a check:
```
if !currentWeekRating.active {
   currentWeekRating.active = true;
   communityUser.userRewards[currentPeriod].activeInCommunity.push(communityId);
}
```

File: contracts/Peeranha.sol (hunk 1 above)
Comment (Freitag-fri) [reply contracts/Peeranha.sol]: in this function we change value
` userContext.userRatingCollection.communityRatingForUser[user].userRewards[rewardPeriod].isPaid = true;`

File: contracts/Peeranha.sol [hunk 14]
```diff
@@ -49,9 +50,13 @@ contract Peeranha is IPeeranha, Initializable {
         SecurityLib.setupRole(userContext, SecurityLib.DEFAULT_ADMIN_ROLE, msg.sender);
     }
 
-    function getRatingToReward(address user, uint16 rewardPeriod) external view override returns(int32) {
-        RewardLib.PeriodRating storage userPeriod = RewardLib.getUserPeriodRating(userContext.userRewards, user, rewardPeriod);
-        return userPeriod.ratingToReward;
+    function getUserRewardCommunities(address user, uint16 rewardPeriod) external override view returns(uint32[] memory) {
+        require(
```
Comment (sergeyi23) [inline contracts/Peeranha.sol]: Why need this check in the view function? I suggest to remove it and just return empty array.

File: contracts/PeeranhaToken.sol [hunk 15]
```diff
@@ -43,15 +44,24 @@ contract PeeranhaToken is ERC20Upgradeable, ERC20PausableUpgradeable, ERC20Cappe
...
+
+    int32 ratingToReward;
+    uint256 tokenReward;
+    for (uint32 i; i < rewardCommunities.length; i++) {
+      ratingToReward = peeranha.getRatingToReward(user, period, rewardCommunities[i]);
+      tokenReward += uint256(ratingToReward) * TokenLib.getRewardCoefficient(); // * 10^18
+    }
     
-    _mint(msg.sender, tokenReward);
+    _mint(user, tokenReward);
```
Comment (sergeyi23) [inline contracts/PeeranhaToken.sol:67]: Add check that `tokenReward > 0`  before calling `_mint`

File: contracts/libraries/UserLib.sol [hunk 16]
```diff
@@ -177,76 +191,91 @@ library UserLib {
...
+
+    int32 newRating = userRating.rating += rating;
+    uint256 pastPeriodsCount = communityUser.rewardPeriods.length;
+
+    RewardLib.UserPeriodRewards storage userPeriodRewards = communityUser.userPeriodRewards[currentPeriod];
+    if (pastPeriodsCount == 0 || communityUser.rewardPeriods[pastPeriodsCount - 1] != currentPeriod) {
+      communityUser.rewardPeriods.push(currentPeriod);
+    }
+
+    if (userPeriodRewards.periodRating[communityId].isActive) {
```
Comment (sergeyi23) [inline contracts/libraries/UserLib.sol]: Should it be `if( !userPeriodRewards.periodRating[communityId].isActive)`?

**Review by sergeyi23 — APPROVED** [review]

**Pull request merged**